rust-src/curve_arithmetic/src/bls12_381_g1hash.rs

```diff
@@ -572,141 +659,163 @@ fn from_coordinates_unchecked(x: Fq, y: Fq, z: Fq) -> Result<G1, CurveDecodingEr
 }
 
-/// Hash to Fq by hashing using Sha512 and decode the first 381 bits as an
-/// unsigned integer in little endian representation. Repeat feeding the
-/// output of the hash into the hash until successful.
-pub fn hash_bytes_to_fq(one: bool, bytes: &[u8]) -> Fq {
-    let mut h = Sha512::new();
-    let mut hash: [u8; 64] = [0u8; 64];
-    h.update(if one { [1u8] } else { [0u8] });
-    h.update(bytes);
-    let mut fqrepr = [0u64; 6];
-    // We need 381 bits to represent the Fq field, so the topmost three bits will
-    // always be unset.
-    let mask: u64 = !(0b111 << 61);
-    loop {
-        hash.copy_from_slice(&h.finalize_reset());
-        for (chunk, f) in hash.chunks_exact(8).zip(fqrepr.iter_mut()) {
-            *f = u64::from_le_bytes(chunk.try_into().expect("Chunk size is always 8."));
-        }
-        // Clear the topmost 3 bits.
-        fqrepr[5] &= mask;
-        // keep trying to hash, until we hit an element in Fq
-        if let Ok(fq) = Fq::from_repr(FqRepr(fqrepr)) {
-            return fq;
-        }
-        h.update(&hash[..]);
+/// Implements https://tools.ietf.org/html/draft-irtf-cfrg-hash-to-curve-10#appendix-G.2.1
+/// Input: u, an element of Fq.
+/// Output: (xn, xd, yn, yd) such that (xn / xd, yn / yd) is a
+///         point on E'.
+#[allow(clippy::many_single_char_names)]
+fn sswu_3mod4(u: Fq) -> (Fq, Fq, Fq, Fq) {
+    let a = Fq::from_repr(FqRepr(E11_A)).unwrap(); // this unwrap can't fail, E11_A is an element of the field
+    let b = Fq::from_repr(FqRepr(E11_B)).unwrap(); // this unwrap can't fail, E11_B is an element of the field
+                                                   // Constants:
+                                                   // 1.  c1 = (q - 3) / 4           # Integer arithmetic
+                                                   // 2.  c2 = sqrt(-Z^3)
+                                                   // Z = 11
+    let z = Fq::from_repr(FqRepr::from(11)).unwrap();
+    // c2 = sqrt(-z^3)
+    let c2 = Fq::from_repr(FqRepr([
+        8011268957077696501,
+        9962099387040634336,
+        8623975380491602827,
+        7731696418485440718,
+        18010667617739806336,
+        276559961644294862,
+    ]))
+    .unwrap();
+
+    // Steps:
+    // 1.  tv1 = u^2
+    let mut tv1 = u;
+    tv1.square();
+
+    // 2.  tv3 = Z * tv1
+    let mut tv3 = z;
+    tv3.mul_assign(&tv1);
+
+    // 3.  tv2 = tv3^2
+    let mut tv2 = tv3;
+    tv2.square();
+
+    // 4.   xd = tv2 + tv3
+    let mut xd = tv2;
+    xd.add_assign(&tv3);
+
+    // 5.  x1n = xd + 1
+    // 6.  x1n = x1n * B
+    let mut x1n = xd;
+    x1n.add_assign(&Fq::one());
+    x1n.mul_assign(&b);
+
+    // 7.   xd = -A * xd
+    let mut neg_a = a;
+    neg_a.negate();
+    xd.mul_assign(&neg_a);
+
+    // 8.   e1 = xd == 0
+    let e1 = xd.is_zero();
+
+    // 9.   xd = CMOV(xd, Z * A, e1)  # If xd == 0, set xd = Z * A
+    if e1 {
+        xd = z;
+        xd.mul_assign(&a); // todo should we use equivalent of CMOV?
```

Comment:
I think for now we should just document that we do not care about whether this function is constant time or not.

The way it is used it does not matter.